### DISCUSSION TITLE
zh/guides/object-counting/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8274] [Docs]
# zh/guides/object-counting/

使用对象计数Ultralytics YOLOv8

https://docs.ultralytics.com/zh/guides/object-counting/

**Comment by HugoSun** [comment]
I want to know how to obtain the in /out count values and return them.

**Reply by glenn-jocher** [reply]
To obtain in/out count values using Ultralytics YOLOv8, you can utilize the `ObjectCounter` from the `ultralytics.solutions` module. Here's a simple example using a video source:

```python
from ultralytics import YOLO
from ultralytics.solutions import object_counter
import cv2

# Load your model
model = YOLO("yolov8n.pt")

# Initialize video capture
cap = cv2.VideoCapture("path/to/video.mp4")
assert cap.isOpened(), "Error reading video file"

# Initialize Object Counter with line points
line_points = [(20, 400), (1080, 400)]
counter = object_counter.ObjectCounter()
counter.set_args(view_img=True, reg_pts=line_points, classes_names=model.names)

# Process video frames
while cap.isOpened():
    success, frame = cap.read()
    if not success:
        break
    tracks = model.track(frame, persist=True)
    frame = counter.start_counting(frame, tracks)
    # Access in/out counts
    in_count = counter.in_count
    out_count = counter.out_count
    print(f"In count: {in_count}, Out count: {out_count}")

cap.release()
cv2.destroyAllWindows()
```

This code will print the in and out counts for the objects being tracked. Adjust the `line_points` to fit the area in your video where you want to count the objects crossing. 🚀

**Comment by scj008** [comment]
I'd like to know if it's possible to set the color and background color of the region text with the parameters, and where I can set the position of the region text, because I don't want it to appear at the very top.

**Reply by glenn-jocher** [reply]
@scj008 hi there! 😊 It's great to see your interest in customizing the appearance and position of the region text in YOLOv8 models. Unfortunately, direct parameters for changing the text color, background color, or position aren't available in the current YOLOv8 implementation.

However, if you're comfortable adjusting the source code, you can modify the visualization functions directly. For drawing the text, look for functions using `cv2.putText()` in the code. You can change the text color by adjusting the color argument, which typically looks like this: `cv2.putText(img, text, org, font, fontScale, color, thickness)`. The `color` argument is where you'll specify the text color, and it's defined as a tuple, e.g., `(255, 255, 255)` for white.

Positioning is determined by the `org` parameter in the same function, which is the bottom-left corner of the text in pixels. Adjusting these values can reposition your text.

For background color, you might need to draw a rectangle behind the text with `cv2.rectangle()` before calling `putText`.

I hope this helps! Let me know if you have any more questions. 🚀

**Reply by scj008** [reply]
I have resolved it in my own way. I modified the def count_labels in your PLOTTING.PY to become:
def count_labels(self, counts=0, count_txt_size=2, color=(255, 255, 255), txt_color=(0, 0, 0), text_x1=None, text_y1=None)

And added the following below:
if text_x1 is not None:
    text_x = text_x1

This allows the use of the passed-in value if available; otherwise, it falls back to your original calculation method. In OBJECT_COUNTER.PY, I added:

if counts_label is not None:
    self.annotator.count_labels(
        counts=counts_label,
        count_txt_size=self.count_txt_thickness,
        txt_color=self.count_txt_color,
        color=self.count_color,
        text_y1=200
        #---
    )

I am not a professional, but I am very interested in your model and have been learning from your work. I am from China and my English is not very good, so I used GPT for many parts of this message. Thank you very much for your response. I am grateful and I may come to learn from you more often in the future.

**Reply by glenn-jocher** [reply]
Hey there! 😄 It's awesome that you've found a solution that works for you. Customizing `count_labels` to include text positioning and color options is a creative approach, and it seems like you've made some valuable adjustments to tailor the functionality to your needs.

Your modifications appear well thought out, and it's great to see you diving into the code to implement the features you need. If you ever feel like sharing improvements or discussing more, the community would surely benefit from your insights. 

Also, don't worry about your English; you're doing great, and using tools to help communicate is a smart way to bridge any gaps. We're here to support each other in learning and making progress. Keep exploring and feel free to reach out anytime you have questions or something cool to share! 🚀

If you're looking into further customization or running into any questions, just drop a message! Happy coding!

**Comment by 1965238126** [comment]
what' the meaning of in count or out count , how to use them  ?

**Reply by glenn-jocher** [reply]
In 和 Out 计数分别指进入和离开定义区域的对象数。您可以通过定义一个特定区域或直线（`reg_pts` 参数），并根据对象跨过该区域或直线的方向来使用它们。例如，您可以设置一个检测门口人数的区域，以区分进入（In count）和离开（Out count）的人数。这对于流量监控、安全跟踪等应用非常有用。

```python
from ultralytics import YOLO
from ultralytics.solutions import object_counter
import cv2

model = YOLO("yolov8n.pt")
cap = cv2.VideoCapture("path/to/video.mp4")

# 设定监控的区域
line_points = [(20, 400), (1080, 400)]  # 直线形式的区域
classes_to_count = [0, 2]  # 人类和汽车类索引

# 初始化对象计数器与视频写入器
counter = object_counter.ObjectCounter()
counter.set_args(view_img=True, reg_pts=line_points, classes_names=model.names, draw_tracks=True)

while cap.isOpened():
    success, im0 = cap.read()
    if not success: break

    # 对跟踪的对象进行计数
    tracks = model.track(im0, persist=True, show=False, classes=classes_to_count)
    im0 = counter.start_counting(im0, tracks)

    # 处理完毕的图像即可视化或存储
    # 示例: cv2.imshow("Result", im0)

cap.release()
```

上述代码示范了如何设置对象计数器，包括如何定义监控区域、针对特定对象类别进行计数，并对结果进行可视化。通过`set_args`方法中的`view_in_counts`和`view_out_counts`参数，您可控制是否显示进入或离开计数。希望这清楚地解答了您的疑问！🙂

**Reply by 1965238126** [reply]
OK . thanks a lot . If the line is drawed , this direction is in count and  at the same time , the other direction is out . But I want to know how  to confirm or  distinguish the in count direction and the  out count direction ?    And   where can I get  the detail class type of classes_to_count in the model = YOLO("yolov8n.pt")  ?  Thinks  a lot .

**Reply by glenn-jocher** [reply]
@1965238126 hey there! 😊 You've touched on an important aspect of object tracking. To distinguish between the 'in' and 'out' count directions when you draw a line, the direction is typically based on the vector orientation of the line. For example, if you draw a line from left to right, objects moving along this vector could be counted as 'in', and those moving against it as 'out'. This can vary based on the implementation but is often configurable.

For the class types available in the `YOLO("yolov8n.pt")` model, you can find them listed in the model's `.names` attribute. Here's a quick code snippet to get those class names:

```python
from ultralytics import YOLO

# Load the model
model = YOLO("yolov8n.pt")

# Print available class names
print(model.names)
```

This will give you a list of all the class types that the model can detect. I hope this helps clarify things for you. Happy coding and tracking! 🚀

**Comment by songziang182** [comment]
Hello, developer, I have a question that is why the track line will become very discontinuous when the parameter persist=True is added and then the track line will be detected. If I don't add this parameter, the track is very smooth, but it gets very bad when I use line counting, so I have to add this parameter, so what can I do to make the track line very smooth and the line detection accurate?

**Reply by glenn-jocher** [reply]
Hello! 😊 It sounds like you're encountering discontinuous tracking lines when using `persist=True`, which helps maintain object IDs across frames, crucial for tasks like line crossing or area counting. This discontinuity might happen due to the tracker temporarily losing the object due to occlusion, rapid movement, or similar challenges, especially in complex scenes.

A smooth tracking line typically indicates a strong, consistent detection and tracking pipeline. To enhance this with `persist=True` for both accurate tracking and line detection, you might consider the following approaches:

1. **Adjust the confidence threshold (`conf`) and Intersection Over Union (IoU) threshold (`iou`)** to fine-tune detection sensitivity and reduce false negatives. For example:

    ```python
    model.track(im0, persist=True, conf=0.4, iou=0.5)
    ```

2. **Experiment with different trackers** specified in the Ultralytics documentation or your configuration (`tracker="bytetrack.yaml"` or `"botsort.yaml"`), as performance can vary based on the scene complexity.

3. **Utilize custom-trained models** that are more closely aligned with your use case, ensuring that the model is as familiar with your specific tracking scenarios as possible.

Implementing these suggestions should help create smoother tracking lines while maintaining the accuracy of the line detection functionality. If adjustments to parameters don't resolve the issue, it could be beneficial to review the video quality or conditions under which tracking becomes challenging and see if any improvements can be made. Happy tracking! 😄

**Comment by 1965238126** [comment]
Dear developer ， 
 My current requirement is:
1. Vehicle count within a fixed time period (such as per minute)
2. At the same time, the number of vehicles passing through this fixed time period
I must use two   speed_obj = speed_estimation.SpeedEstimator()  and  counter = object_counter.ObjectCounter() at the same time. But this count is cumulative 。Is there a simpler way ？
Could you please give me a Python code example  ？ 
3.  different reg_pts  setting results different speed estimation result  .  
 width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH)/2)  
 height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
 line_pts2  = [(0,width),(height,width)]
 speed_obj = speed_estimation.SpeedEstimator()
  speed_obj.set_args(reg_pts=line_pts2,names=names,view_img=False)   
I got the max speed in my cap (mp4 file )  over 110 KM /H . But  I know  that this is impossible. Do you have any best practice on how to set this parameter?
Thanks a lot .

**Reply by pderrenger** [reply]
@1965238126 hey there! 👋

For your vehicle counting requirement, using both `speed_estimation.SpeedEstimator()` and `object_counter.ObjectCounter()` is a great approach! If you want to reset the count at the end of each time period (like every minute), you can simply reset your `counter` object inside a loop that runs based on your fixed time period. Here's a compact way to manage this:

```python
import time
from ultralytics import YOLO, speed_estimation, object_counter

# Load your model
model = YOLO('yolov8n.pt')

# Initialize your objects
speed_obj = speed_estimation.SpeedEstimator()
counter = object_counter.ObjectCounter()

# Assuming you have a function to process video frames
def process_frame(frame):
    ...
    # Use speed_obj and counter here
    ...

# Main loop
start_time = time.time()
while True:  # Your condition to run
    frame = ... # Capture frame
    process_frame(frame)
    
    # Check if a minute has passed
    if time.time() - start_time > 60:
        # Reset counter
        counter.reset()
        start_time = time.time()
```

Regarding the speed estimation and `reg_pts` setting, it's vital to accurately set this based on your scene. The placement of `line_pts2` sounds about right, but ensuring the precision of your measurements (like camera height and angle) is key to accurate speed estimation. If you're facing inaccurately high speeds, consider reviewing your setup or experimenting with different `reg_pts` positions. There's no one-size-fits-all since environments vary greatly, but starting with points that span the width and height where vehicles pass most frequently is a good practice.

Hope this helps! Feel free to share more about your setup if you need further assistance. 😊

**Reply by 1965238126** [reply]
1 .    In my class ObjectCounter , this is no reset funcation . 
2.    In my  class SpeedEstimator  , I saw these coding :    
     if self.time_difference > 0:

                dist_difference = np.abs(track[-1][1] - self.trk_previous_points[trk_id][1])

                speed = dist_difference   /  self.time_difference

                speed_time = "speed : " +  str(int(speed))  + " , time:"   +  str(int(dist_difference))

                self.dist_data[trk_id] =  speed_time
      I think the  unit  of  dist_difference  is  pixels and the unit of time_difference is second . How to  converted  the piexls/second  to kilometers/hour 。

3.  In my code : 

def show_frame(source):
   ...
    while cap.isOpened():

            ret, im0 = cap.read()
            ...
                     
            if frame_count % frames_to_skip == 0:
                # count_traff_time()
             
                  tracks_speed = model.track(im0, persist=False, show=False)  # speed
                  im0_speed = speed_obj.estimate_speed(im0, tracks_speed)
                  ana_data = speed_obj.dist_data
                   ...
            frame_count += 1
    ## end 
  
    When I estimate the speed , some frames of the video have been  skipped  .  I want to know  if  this skipping  affect  the accuracy of the estimation of speed? If so, how can it be corrected .

I am sorry  , I'm a novice in this area. Thank you.

**Reply by pderrenger** [reply]
Hey there! 👋 It looks like you're making great progress with your object counting and speed estimation projects! Let me address your queries:

1. **Reset Function for ObjectCounter:** It seems you need a way to reset the count. If `ObjectCounter` does not have a `reset()` method built-in, a simple workaround could be to re-initialize the `counter` object at the interval you desire, e.g., every minute for a fresh count.

2. **Pixels/Second to Kilometers/Hour Conversion:** Indeed, the speed estimation in pixels/second needs to be converted to a physical unit like km/h. This requires knowing the scale (how many pixels correspond to a real-world distance). For example, if 1 pixel corresponds to 0.05 meters, and assuming `speed` is in pixels/second, convert it as follows:
   
   ```python
   speed_km_h = (speed * 0.05 * 3600) / 1000  # Convert pixels/second to km/h
   ```

   Adjust `0.05` according to your scale.

3. **Skipping Frames Impact on Speed Estimation:** Skipping frames can indeed affect the accuracy, as it might skip important motions leading to miscalculations of speed due to larger perceived distances between detections in consecutive processed frames. A potential correction is to adjust the `time_difference` to account for the number of skipped frames, essentially considering the real-time gap between the processed frames.

No need to apologize for being new to this area, we all start somewhere, and it looks like you're on the right path! If you need further assistance or have more questions, don't hesitate to ask. Happy coding! 😊
